### PR TITLE
Add SMW 4 support (but with failing tests)

### PR DESCRIPTION
**Pull request by malberts**
Based on https://github.com/NaimKabir/semantic-mediawiki/tree/kabir/upgrade-to-smw-4.0.0 (https://github.com/NaimKabir/semantic-mediawiki/pull/2)

Additional changes:
* add environment variable used by tests (MW_INSTALL_PATH)
* new installation requirement: `wfLoadExtension()` (https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/INSTALL.md#step-3)
* use `test` command, because `phpunit` has the `--testdox` flag which is overly verbose IMO

The tests run now, but 2 of the sets fail:
**semantic-mediawiki-unit**: 
```
=== RUN: Command Test: semantic-mediawiki-unit
--- FAIL
duration: 12.974125401s
stdout: 
Semantic MediaWiki: 4.0.0, git: 
                    SMWSQLStore (sqlite:3.34.1)

MediaWiki:          1.37.1, git: n/a
                    MediaWiki vendor autoloader

Site language:      en
Execution time:     2022-02-12 12:00

Debug logs:         Disabled
Xdebug:             Disabled (or not installed)

Intl/ICU:           7.4.27 / 67.1
PCRE:               10.35 2020-05-09


Fatal error: Uncaught UnexpectedValueException: Got connection to 'my_wiki', but expected local domain ('my_wiki-sunittest_') in /var/www/html/includes/libs/rdbms/loadbalancer/LoadBalancer.php:1198
Stack trace:
#0 /var/www/html/includes/libs/rdbms/loadbalancer/LoadBalancer.php(986): Wikimedia\Rdbms\LoadBalancer->getLocalConnection(0, 4)
#1 /var/www/html/includes/libs/rdbms/loadbalancer/LoadBalancer.php(963): Wikimedia\Rdbms\LoadBalancer->getServerConnection(0, 'my_wiki-sunitte...', 4)
#2 /var/www/html/includes/libs/rdbms/loadbalancer/LoadBalancer.php(1096): Wikimedia\Rdbms\LoadBalancer->getConnection(-2, Array, 'my_wiki-sunitte...', 4)
#3 /var/www/html/extensions/SemanticMediaWiki/src/MediaWiki/Connection/LoadBalancerConnectionProvider.php(101): Wikimedia\Rdbms\LoadBalancer->getConnectionRef(-2, Array, 'my_wiki-sunitte...')
#4 /var/www/html/extensions/SemanticMediaWiki/src/Connection/ConnRef.php(60): SMW\MediaWiki\Connection\LoadBalancerConnectionProvider->getConnection()
#5 /var/www/html/extensions/SemanticMediaWiki/src/ in /var/www/html/includes/libs/rdbms/loadbalancer/LoadBalancer.php on line 1198

stderr: > php ${MW_INSTALL_PATH}/tests/phpunit/phpunit.php -c phpunit.xml.dist '--testsuite' 'semantic-mediawiki-unit' '--filter' '^((?!ScopeMemoryLimiterTest::testExecute|CsvFileExportPrinterTest::testLink|ProtectionValidatorTest::testIsClassifiedAsImportPerformerProtected_CreatorAndCurrentUserDontMatch|TemplateFileExportPrinterTest::testLink).)*$'
Warning: Use of undefined constant SemanticMediaWiki - assumed 'SemanticMediaWiki' (this will throw an Error in a future version of PHP) in /var/www/html/LocalSettings.php on line 174
Using PHP 7.4.27
PHPUnit 8.5.23 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.27
Configuration: /var/www/html/extensions/SemanticMediaWiki/phpunit.xml.dist

...............................................................  63 / 506 ( 12%)
............................................................... 126 / 506 ( 24%)
............................................................... 189 / 506 ( 37%)
............................................................... 252 / 506 ( 49%)
..............................S................................ 315 / 506 ( 62%)
............................................................... 378 / 506 ( 74%)
............................................................... 441 / 506 ( 87%)
............................................................... 504 / 506 ( 99%)
..                                                              506 / 506 (100%)

Time: 11.75 seconds, Memory: 94.50 MB

There was 1 skipped test:

1) SMW\Tests\DITimeTest::testDateTimeWithLargeMs
For some reason this started to fail on 5.6.19 (worked on 5.6.18)

/var/www/html/extensions/SemanticMediaWiki/tests/phpunit/includes/dataitems/DITimeTest.php:90

OK, but incomplete, skipped, or risky tests!
Tests: 506, Assertions: 1158, Skipped: 1.
Script php ${MW_INSTALL_PATH}/tests/phpunit/phpunit.php -c phpunit.xml.dist handling the test event returned with error code 255

Error: Test 'semantic-mediawiki-unit' exited with incorrect error code. Expected: 0, Actual: 255
```

**semantic-mediawiki-integration**
Times out after 300 seconds, even when disabling all the other test sets.

### CODE REVIEW

File: templates/container/BUILD.bazel.jinja [hunk 1]
```diff
@@ -46,6 +46,7 @@ container_run_and_commit(
 	"php maintenance/install.php my_wiki admin --dbname my_wiki --pass a-nice-password --dbtype sqlite --server 'http://localhost:8080' --scriptpath '' ",
 	"chown www-data:www-data data data/*",
 	# Enable Semantic MediaWiki and then update MediaWiki to deal with the upgrade
+	"echo 'wfLoadExtension( \"SemanticMediaWiki\" );' >> LocalSettings.php",
```
Comment (NaimKabir) [inline templates/container/BUILD.bazel.jinja]: Here's a modification which will get this to register as a PHP string (right now it's treated as an unidentified constant that is converted to a string—but this is brittle behavior!).

```suggestion
	# We have to use funky PHP heredoc syntax to insert a string, given our string format
	# limitations in Bazel. :(
	"echo 'wfLoadExtension(<<<END' >> LocalSettings.php", 
	"echo 'SemanticMediaWiki' >> LocalSettings.php", 
	"echo 'END);' >> LocalSettings.php", 
```

Comment (NaimKabir) [reply templates/container/BUILD.bazel.jinja]: After this is committed we have to run `./update_versions.py` again, though.

**Review by NaimKabir — CHANGES_REQUESTED** [review]
Hey @malberts — thank you so much for this contribution.

The failing tests are certainly a problem. How about we just remove the failing test suites entirely in this PR, since I suspect tests are failing on SMW master given my [previous conversations with their team](https://twitter.com/KabirCreates/status/1433653123537584128?s=20&t=yjGSkcL3ldlb_SaqvfDpsQ). 

Alernatively, you can add to the GREP rules that exclude the specific failing tests, if you're willing to put in that tedious effort.

To be safe while still getting this change in, how about we post a couple of screenshots of a working MediaWiki install? That would be sufficient to merge this in.

So that would be:

1. Run `bazel run //container:test-container`
2. Run `docker run --name test -p 8080:80 -d bazel/container:test-container`
3. Open a browser to `localhost:8080` and verify a working install with [these steps](https://www.semantic-mediawiki.org/wiki/Help:Verify_the_installation)
4. Post some screenshots in this PR

Thanks again for your time, here

**Comment by malberts** [comment]
Weird escaping behaviour from Bazel there. I added your suggestion and commented out the failing test sets. I'm not actually using this image so I cannot spend too much time on it (I just remembered you posted on the SMW mailing list and thought this might be a simple fix). And yes, some of the tests are failing upstream for 1.37 anyway, e.g.:
https://github.com/SemanticMediaWiki/SemanticMediaWiki/runs/5166665676?check_suite_focus=true#step:10:15928

Screenshots from the steps:
Screenshot_20220213_114545

Screenshot_20220213_114639

Screenshot_20220213_114703

Screenshot_20220213_114743

**Comment by NaimKabir** [comment]
Excellent, thank you—I'll merge this (and then in a follow-up I'll amend the Jinja to also remove the failing tests, we can always add them back later).

I appreciate you taking the time to help out

**Review by NaimKabir — APPROVED** [review]

**Pull request merged**